File: LeetCode/DP/Golang/Leetcode688-骑士在棋盘的概率.go
package sbgo

func knightProbability(n int, k int, row int, column int) float64 {
	f := make([][][]float64, n)
	for i := 0; i < n; i++ {
		f[i] = make([][]float64, n)
		for j := 0; j < n; j++ {
			f[i][j] = make([]float64, k+1)
		}
	}
	dx, dy := []int{-2, -1, 1, 2, 2, 1, -1, -2}, []int{1, 2, 2, 1, -1, -2, -2, -1}
	for st := 0; st <= k; st++ {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if st == 0 {
					f[i][j][0] = 1
				} else {
					for d := 0; d < 8; d++ {
						ni, nj := i+dx[d], j+dy[d]
						if ni < 0 || ni >= n || nj < 0 || nj >= n {
							continue
						}
						f[i][j][st] += f[ni][nj][st-1] / 8
					}
				}
			}
		}
	}
	return f[row][column][k]
}

File: LeetCode/DP/JS/Leetcode688-骑士在棋盘的概率.js
var knightProbability = function(n, k, row, column) {
    const dx = [-2, -1, 1, 2, 2, 1, -1, -2],dy = [1, 2, 2, 1, -1, -2, -2, -1];
    const f = new Array(n).fill(0).map(() => new Array(n).fill(0).map(() => new Array(k + 1).fill(0)));
    for(let st = 0;st <= k;st++) {
        for(let i = 0;i < n;i++){
            for(let j = 0;j < n;j++) {
                if(st == 0) f[i][j][st] = 1;
                else{
                    for(let d = 0;d < 8;d++){
                        let ni = i + dx[d],nj = j + dy[d];
                        if(ni < 0 || ni >= n || nj < 0 || nj >= n) continue;
                        f[i][j][st] += f[ni][nj][st - 1] / 8;
                    }
                }
            }
        } 
    }
    return f[row][column][k];
};